bam/reader.go

```diff
@@ -101,98 +99,121 @@ const (
 	AllVariableLengthData        // Omit sequence, quality and auxiliary data.
 )
 
-// Read returns the next sam.Record in the BAM stream. The sam.Record
-// returned will not contain the sequence, quality or auxiliary tag
-// data if CoreOnly(true) has been called.
-func (br *Reader) Read() (*sam.Record, error) {
-	if br.c != nil && vOffset(br.r.LastChunk().End) >= vOffset(br.c.End) {
-		return nil, io.EOF
-	}
+// buffer is light-weight read buffer.
+type buffer struct {
+	off  int
+	data []byte
+}
 
-	r := errReader{r: br.r}
-	bin := binaryReader{r: &r}
+func (b *buffer) bytes(n int) []byte {
+	s := b.off
+	b.off += n
+	return b.data[s:b.off]
+}
 
-	// Read record header data.
-	blockSize := int(bin.readInt32())
-	r.n = 0 // The blocksize field is not included in the blocksize.
+func (b *buffer) len() int {
+	return len(b.data) - b.off
+}
+
+func (b *buffer) discard(n int) {
+	b.off += n
+}
+
+func (b *buffer) readUint8() uint8 {
+	b.off++
+	return b.data[b.off-1]
+}
+
+func (b *buffer) readUint16() uint16 {
+	return binary.LittleEndian.Uint16(b.bytes(2))
+}
+
+func (b *buffer) readInt32() int32 {
+	return int32(binary.LittleEndian.Uint32(b.bytes(4)))
+}
 
+func (b *buffer) readUint32() uint32 {
+	return binary.LittleEndian.Uint32(b.bytes(4))
+}
+
+// newBuffer returns a new buffer reading from the Reader's underlying bgzf.Reader and
+// updates the Reader's lastChunk field.
+func newBuffer(br *Reader) (*buffer, error) {
+	var buf [4]byte
+	n, err := io.ReadFull(br.r, buf[:4])
+	if err != nil {
+		return nil, err
+	}
+	if n != 4 {
+		return nil, errors.New("bam: invalid record: short block size")
+	}
+	b := &buffer{data: buf[:4]}
+	size := int(b.readInt32())
+	b.off, b.data = 0, make([]byte, size)
 	// br.r.Chunk() is only valid after the call the Read(), so this
 	// must come after the first read in the record.
 	tx := br.r.Begin()
```

Comment:
I made the changes except this one. Not sure what you mean. The tx has to span the Read() of the record and we want to check for errors before starting the tx, right?